simulation/traffic_simulator/src/entity/entity_base.cpp

```diff
@@ -713,5 +713,147 @@ auto EntityBase::updateTraveledDistance(const double step_time) -> double
+  if (reachPosition(entity_target, threshold)) {
+    if (this->getStatus().getTwist().linear.x < accel_limit * loop_period / 1000) {
+    } else {
+      RCLCPP_WARN_ONCE(
+        rclcpp::get_logger("traffic_simulator"),
+        "The entity has already arrived to the target lanelet and the entity is not nearly "
+        "stopped.");
+    }
+    target_speed_ = 0;
+    return true;
+  }
+
+  job_list_.append(
+    [this, ego_target, entity_target, accel_limit, threshold, loop_period](double) {
+      auto entity_distance = getDistanceToTargetLaneletPose(entity_target);
+      auto ego_distance = this->hdmap_utils_ptr_->getLongitudinalDistance(
+        static_cast<LaneletPose>(other_status_.find("ego")->second.getLaneletPose()),
+        static_cast<LaneletPose>(ego_target));
+
+      if (!ego_distance.has_value()) {
+        RCLCPP_WARN_ONCE(
+          rclcpp::get_logger("traffic_simulator"),
+          "Failed to get distance between Ego and target lanelet pose. Check if Ego has already "
+          "passed the target lanelet. If not, please contact the developer since there might be an "
+          "undiscovered bug.");
+        return true;
+      }
+      const auto ego_velocity = other_status_.find("ego")->second.getTwist().linear.x;
+      const auto entity_velocity = this->getStatus().getTwist().linear.x;
+      const auto ego_arrival_time =
+        (std::abs(ego_velocity) > std::numeric_limits<double>::epsilon())
+          ? ego_distance.value() / ego_velocity
+          : 0;
+
+      auto entity_velocity_to_synchronize = entity_velocity;
+
+      const auto accel_limit_ = accel_limit;
+      const auto distance_margin = threshold;
+
+      if (entity_velocity > accel_limit * ego_arrival_time) {
+        ///@brief Making entity slow down since the speed is too fast
+        entity_velocity_to_synchronize = entity_velocity - accel_limit_ * loop_period / 1000;
+      } else if (
+        entity_velocity * ego_arrival_time - entity_velocity * entity_velocity / (accel_limit * 2) <
+        entity_distance - distance_margin) {
+        ///@brief Making entity speed up
+        entity_velocity_to_synchronize = entity_velocity + accel_limit_ * loop_period / 1000;
+      } else if (
+        entity_velocity * ego_arrival_time - entity_velocity * entity_velocity / (accel_limit * 2) >
+        entity_distance - distance_margin) {
+        ///@brief Making entity slow down
+        entity_velocity_to_synchronize = entity_velocity - accel_limit_ * loop_period / 1000;
+      } else {
+        ///@brief Making entity keep the current speed
+        entity_velocity_to_synchronize = entity_velocity;
+      }
+      /**
+       * @warning using this->requestSpeedChange here does not work in some kind of reason.
+       * It seems that after this function is called by some reason, func_on_cleanup will be deleted and become nullptr
+       */
+      target_speed_ = entity_velocity_to_synchronize;
+      return false;
+    },
```

Comment:
Logic is too complex, please consider using lambda function.
The scope of temporary variables should be as narrow as possible.